pkg/relayer/proxy/http_server.go

```diff
@@ -0,0 +1,146 @@
+// one, such as QUIC or pure TCP.
+func NewHTTPServer(
+	logger polylog.Logger,
+	serverConfig *config.RelayMinerServerConfig,
+	servedRelaysProducer chan<- *types.Relay,
+	relayAuthenticator relayer.RelayAuthenticator,
+	relayMeter relayer.RelayMeter,
+	blockClient client.BlockClient,
+	sharedQueryClient client.SharedQueryClient,
+	sessionQueryClient client.SessionQueryClient,
+) relayer.RelayServer {
+	return &relayMinerHTTPServer{
+		logger:               logger,
+		server:               &http.Server{Addr: serverConfig.ListenAddress},
+		relayAuthenticator:   relayAuthenticator,
+		servedRelaysProducer: servedRelaysProducer,
+		serverConfig:         serverConfig,
+		relayMeter:           relayMeter,
+		blockClient:          blockClient,
+		sharedQueryClient:    sharedQueryClient,
+		sessionQueryClient:   sessionQueryClient,
+	}
+}
+
+// Start starts the service server and returns an error if it fails.
+// It also waits for the passed in context to end before shutting down.
+// This method is blocking and should be called in a goroutine.
+func (server *relayMinerHTTPServer) Start(ctx context.Context) error {
+	go func() {
+		<-ctx.Done()
+		_ = server.server.Shutdown(ctx)
+	}()
+
+	// Set the HTTP handler.
+	server.server.Handler = server
+
+	return server.server.ListenAndServe()
+}
+
+// Stop terminates the service server and returns an error if it fails.
+func (server *relayMinerHTTPServer) Stop(ctx context.Context) error {
+	return server.server.Shutdown(ctx)
+}
+
+// ServeHTTP listens for incoming relay requests. It implements the respective
+// method of the http.Handler interface. It is called by http.ListenAndServe()
+// when relayMinerHTTPServer is used as an http.Handler with an http.Server.
+// (see https://pkg.go.dev/net/http#Handler)
+func (server *relayMinerHTTPServer) ServeHTTP(writer http.ResponseWriter, request *http.Request) {
+	ctx := request.Context()
+
+	// Determine whether the request is upgrading to websocket.
+	if isWebSocketRequest(request) {
+		server.logger.Debug().Msg("detected asynchronous relay request")
+
+		if err := server.handleAsyncConnection(ctx, writer, request); err != nil {
+			// Reply with an error if the relay could not be served.
+			server.replyWithError(err, nil, writer)
+			server.logger.Warn().Err(err).Msg("failed serving asynchronous relay request")
+			return
+		}
+	} else {
+		server.logger.Debug().Msg("detected synchronous relay request")
+
+		if relayRequest, err := server.serveSyncRequest(ctx, writer, request); err != nil {
+			// Reply with an error if the relay could not be served.
+			server.replyWithError(err, relayRequest, writer)
+			server.logger.Warn().Err(err).Msg("failed serving synchronous relay request")
+			return
+		}
+	}
+}
+
+// isWebSocketRequest checks if the request is trying to upgrade to WebSocket.
+func isWebSocketRequest(r *http.Request) bool {
+	// Check if the request is trying to upgradeHeader to WebSocket
```

Comment:
Can you link to something (form ietf?) explaining the Upgrade + Connection + Websocket strings?